File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2017/Y17D09.kt
package org.gristle.pdxpuzzles.advent.y2017

import org.gristle.pdxpuzzles.advent.utilities.Day

class Y17D09(private val input: String) : Day {

    val solution: Pair<Int, Int> by lazy {
        var inGarbage = false
        var garbage = 0
        var depth = 0
        var score = 0
        var last = ' '
        input.forEach { c ->
            if (inGarbage) {
                if (c == '>' && last != '!') inGarbage = false
                if (!(c in "!>" || last == '!')) garbage++
                last = if (last == '!') ' ' else c
            } else {
                when (c) {
                    '<' -> inGarbage = true

                    '{' -> {
                        depth++
                        score += depth
                    }

                    '}' -> {
                        depth--
                    }
                }
            }
        }
        score to garbage
    }

    override fun part1() = solution.first

    override fun part2() = solution.second
}


fun main() = Day.runDay(Y17D09::class)

//    Class creation: 2ms
//    Part 1: 9251 (2ms)
//    Part 2: 4322 (0ms)
//    Total time: 5ms

File: rust/advent/src/bin/2017/y17d09.rs
use advent::utilities::get_input::get_input;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input = (usize, usize);
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(17, 9).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    let mut in_garbage = false;
    let mut garbage = 0;
    let mut depth = 0;
    let mut score = 0;
    let mut last = b' ';
    for &b in input.as_bytes() {
        if in_garbage {
            if b == b'>' && last != b'!' {
                in_garbage = false;
            }
            if !(b == b'!' || b == b'>' || last == b'!') {
                garbage += 1;
            }
            last = if last == b'!' { b' ' } else { b };
        } else {
            match b {
                b'<' => {
                    in_garbage = true;
                }
                b'{' => {
                    depth += 1;
                    score += depth;
                }
                b'}' => {
                    depth -= 1;
                }
                _ => {}
            }
        }
    }
    (score, garbage)
}

fn part1(input: &Input) -> Output {
    input.0
}

fn part2(input: &Input) -> Output {
    input.1
}

#[test]
fn default() {
    let input = get_input(17, 9).unwrap();
    let input = parse_input(&input);
    assert_eq!(9251, part1(&input));
    assert_eq!(4322, part2(&input));
}

// Input parsed (219μs)
// 1. 9251 (4μs)
// 2. 4322 (1μs)
// Total: 227μs
